src/lualib/Promise.ts

```diff
@@ -0,0 +1,185 @@
+/* eslint-disable @typescript-eslint/promise-function-async */
+
+// Promises implemented based on https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise
+// and https://promisesaplus.com/
+
+enum __TS__PromiseState {
+    Pending,
+    Fulfilled,
+    Rejected,
+}
+
+type FulfillCallback<TData, TResult> = (value: TData) => TResult | PromiseLike<TResult>;
+type RejectCallback<TResult> = (reason: any) => TResult | PromiseLike<TResult>;
+
+function __TS__PromiseDeferred<T>() {
+    let resolve: FulfillCallback<T, unknown>;
+    let reject: RejectCallback<unknown>;
+    const promise = new Promise<T>((res, rej) => {
+        resolve = res;
+        reject = rej;
+    });
+
+    return { promise, resolve, reject };
+}
+
+function __TS__IsPromiseLike<T>(thing: unknown): thing is PromiseLike<T> {
+    return thing instanceof __TS__Promise;
+}
+
+class __TS__Promise<T> implements Promise<T> {
+    public state = __TS__PromiseState.Pending;
+    public value?: T;
+    public rejectionReason?: any;
+
+    private fulfilledCallbacks: Array<FulfillCallback<T, unknown>> = [];
+    private rejectedCallbacks: Array<RejectCallback<unknown>> = [];
+    private finallyCallbacks: Array<() => void> = [];
+
+    public [Symbol.toStringTag]: string; // Required to implement interface, no output Lua
+
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/resolve
+    public static resolve<TData>(this: void, data: TData): Promise<TData> {
+        // Create and return a promise instance that is already resolved
+        const promise = new __TS__Promise<TData>(() => {});
+        promise.state = __TS__PromiseState.Fulfilled;
+        promise.value = data;
+        return promise;
+    }
+
+    // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/reject
+    public static reject(this: void, reason: any): Promise<never> {
+        // Create and return a promise instance that is already rejected
+        const promise = new __TS__Promise<never>(() => {});
+        promise.state = __TS__PromiseState.Rejected;
+        promise.rejectionReason = reason;
+        return promise;
+    }
+
+    constructor(executor: (resolve: (data: T) => void, reject: (reason: any) => void) => void) {
+        executor(this.resolve.bind(this), this.reject.bind(this));
```

Comment:
This should be wrapped in a try catch if the executor fails the promise should be rejected.
```suggestion
try {
        executor(this.resolve.bind(this), this.reject.bind(this));
} catch (error) {
        reject(error)
}
```

Reply:
Probably also good idea to add a test for that, haven't looked at those yet.

Reply:
Good catch, I missed this